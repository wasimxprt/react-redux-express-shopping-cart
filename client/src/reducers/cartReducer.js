import * as actions from "../constants/actionTypes";

export const initialState = {
    cart: [],
    totalAmount: 0
};

export const cartReducer = (state = initialState, action) => {

    let cart = state.cart;

    switch (action.type) {

        case 'ADD_TO_CART':

            cart.push(action.payload);
            const total = Object.values(cart)
                .reduce((result, cartItem) => result + cartItem.product.price * cartItem.quantity, 0);


            return {
                ...state,
                cart: cart,
                totalAmount: total
            };

        case 'INCREASE_QUANTITY':
            {
                let item = cart.find(item => item.product._id == action.payload.productId);

                var filtered = cart.filter(function (item) { return item.product._id != action.payload.productId; });
                item.quantity += 1;
                filtered.push(item);
                filtered.sort((a, b) => a.product.name.localeCompare(b.product.name));

                const total = Object.values(filtered)
                    .reduce((result, cartItem) => result + cartItem.product.price * cartItem.quantity, 0);

                return {
                    ...state,
                    cart: filtered,
                    totalAmount: total
                };
            }


        case 'DECREASE_QUANTITY':
            {
                let item = cart.find(item => item.product._id == action.payload.productId);

                if (item.quantity === 1) {
                    return {
                        ...state,
                        cart: cart.filter(item => item.product._id != action.payload.productId)
                    }
                }

                var filtered = cart.filter(function (item) { return item.product._id != action.payload.productId; });
                item.quantity -= 1;
                filtered.push(item);
                filtered.sort((a, b) => a.product.name.localeCompare(b.product.name));

                const total = Object.values(filtered)
                    .reduce((result, cartItem) => result + cartItem.product.price * cartItem.quantity, 0);

                return {
                    ...state,
                    cart: filtered,
                    totalAmount: total
                };
            }

        case 'REMOVE_FROM_CART':
            {
                let newCart = cart.filter(item => item.product._id != action.payload.productId);
                const total = Object.values(newCart)
                    .reduce((result, cartItem) => result + cartItem.product.price * cartItem.quantity, 0);

                return {
                    ...state,
                    cart: newCart,
                    totalAmount: total
                };
            }

        default:
            return state;
    }
};

export default cartReducer;