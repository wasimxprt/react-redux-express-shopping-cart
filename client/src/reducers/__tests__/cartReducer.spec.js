import cartReducer from '../cartReducer';
import initialState from '../cartReducer';

describe('cart reducer', () => {
    it('should return the initial state', () => {
        expect(cartReducer(undefined, {})).toMatchSnapshot()
    })

    it('should handle ADD_TO_CART', () => {
        expect(
            cartReducer({
                
                totalAmount: 0
            },
                {
                    type: 'ADD_TO_CART',
                    payload: {
                        product:{_id: "5f1d397ea15c6e1be4aa10e2",name: "Handcrafted Concrete Shoes",price: 999},
                        quantity: 1
                    }
                })
        ).toMatchSnapshot()
    })
})