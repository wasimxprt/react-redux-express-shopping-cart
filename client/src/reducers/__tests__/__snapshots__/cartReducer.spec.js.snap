// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cart reducer should handle ADD_TO_CART 1`] = `
Object {
  "cart": Array [
    Object {
      "product": Object {
        "_id": "5f1d397ea15c6e1be4aa10e2",
        "name": "Handcrafted Concrete Shoes",
        "price": 999,
      },
      "quantity": 1,
    },
  ],
  "totalAmount": 999,
}
`;

exports[`cart reducer should return the initial state 1`] = `
Object {
  "cart": Array [],
  "totalAmount": 0,
}
`;
